libsolidity/formal/SMTChecker.cpp

```diff
@@ -267,15 +278,15 @@ void SMTChecker::endVisit(FunctionCall const& _funCall)
 	{
 		solAssert(args.size() == 1, "");
 		solAssert(args[0]->annotation().type->category() == Type::Category::Bool, "");
-		checkCondition(!(expr(*args[0])), _funCall.location(), "Assertion violation");
-		m_interface->addAssertion(expr(*args[0]));
+		checkCondition(getCurrentPathConditions() && !(expr(*args[0])), _funCall.location(), "Assertion violation");
 	}
 	else if (funType.kind() == FunctionType::Kind::Require)
 	{
 		solAssert(args.size() == 1, "");
 		solAssert(args[0]->annotation().type->category() == Type::Category::Bool, "");
-		m_interface->addAssertion(expr(*args[0]));
-		checkCondition(!(expr(*args[0])), _funCall.location(), "Unreachable code");
+		checkCondition(getCurrentPathConditions() && expr(*args[0]), _funCall.location(), "Unreachable code", smt::CheckResult::UNSATISFIABLE);
+		smt::Expression impl = smt::Expression::implies(getCurrentPathConditions(), expr(*args[0]));
+		m_interface->addAssertion(impl);
```

Comment:
Shouldn't this rather call `pushPathCondition`?

Reply:
The idea here is that the condition inside the `require` is true whenever the current conditions are, so by using this implication the require condition is valid in the same scope as potential `ite` variables, and therefore the require condition doesn't have to be pushed.
Example:
```
uint x;
if(a == 0) {
    require(b == 0);
    x = b + 1;
} else {
    require(b == 1);
    x = b + 2;
}
assert(x == 1 || x == 3);
```
leads to the constraints
```
- (a_0 == 0) -> (b_0 == 0)
- (a_0 != 0) -> (b_0 == 1)
- x_1 = ite(a_0 == 0, (b_0 + 1), (b_0 + 2))
```
So if `(a_0 == 0)` is `true`, the constraint `(b_0 == 0)` is true and will be implicitly taken into account by the solver whenever `x_1` is used (same for the negative case).

Reply:
Ah, I see! Still, shouldn't we treat all conditions that affect control-flow in the same way?

Reply:
ok, in combination with the discussion below, yours should be the right approach :-)